rs/messaging/src/routing/stream_handler.rs

```diff
@@ -766,9 +763,27 @@ impl StreamHandlerImpl {
                     *available_guaranteed_response_memory -=
                         stream.push(generate_reject_response(&request, code, err.to_string()))
                             as i64;
-                    stream.push_accept_signal()
+                    stream.push_accept_signal();
+                }
+                Some((err, RequestOrResponse::Request(_))) => {
+                    // Unable to induct a request, push a reject signal.
+                    let reason = match err {
+                        StateError::CanisterMigrating { .. } => RejectReason::CanisterMigrating,
+                        StateError::CanisterNotFound(_) => RejectReason::CanisterNotFound,
+                        StateError::CanisterStopped(_) => RejectReason::CanisterStopped,
+                        StateError::CanisterStopping(_) => RejectReason::CanisterStopping,
+                        StateError::QueueFull { .. } => RejectReason::QueueFull,
+                        StateError::OutOfMemory { .. } => RejectReason::OutOfMemory,
+                        StateError::NonMatchingResponse { .. }
+                        | StateError::BitcoinNonMatchingResponse { .. } => RejectReason::Unknown,
+                    };
+                    stream.push_reject_signal(reason);
+                }
+                Some((StateError::CanisterMigrating { .. }, RequestOrResponse::Response(_))) => {
+                    // Unable to deliver a response due to migrating canister, push reject signal.
+                    stream.push_reject_signal(RejectReason::CanisterMigrating);
                 }
-                Err((_, RequestOrResponse::Response(_))) => {
+                Some((_, RequestOrResponse::Response(_))) => {
                     unreachable!("No signals are generated for response induction failures except for CanisterMigrating");
```

Comment:
I would handle both unreachable cases (this and the one above) consistently: either both `unreachable!()` or both returning `RejectReason::Unknown`.

Reply:
We don't want to panic for requests because the error is just passed along from `ReplicatedState::push_input()`; for responses the code in `induct_message_impl()` is such that only `StateError::CanisterMigrating { .. }` can come up so it would have to be a local change that triggers this. You wouldn't get away with it without changing unit tests which is why I felt it's ok to panic and we are not supposed to generate reject signals for responses for anything other than canister migrating.

OTOH we have to handle all combinations of response and `RejectSignal` anyway so we could also just treat all messages the same and remove the special cases for responses. If anything other than canister migrating gets through (it's not possible from the code as it is now though), `handle_rejected_messages()` on the other subnet will log an error so it won't go unnoticed.

I'm fine with either since the panic can't trigger on mainnet since it won't pass the CI if there is an issue. Then again perhaps we shouldn't rely on it if we don't have to. WDYT @derlerd-dfinity ?

Reply:
Our policy is not to crash if there is a meaningful way to go ahead. So if returning `RejectReason::Unknown` is an option that works this is the way to go. 

Lets maybe discuss this in more detail. Perhaps also @oggy-dfin has an opinion?

Reply:
Note that there is no code path into this `unreachable!()`. It is literally unreachable. The only combination of `(RejectReason, RequestOrResponse::Response)` `induct_message_impl()` returns is `(RejectReason::CanisterMigrating, RequestOrResponse::Response)` since for all other combinations the message is silently dropped. It is more of a documentation of this behavior I suppose than an actually functionality.

Reply:
Yeah, I saw it as well. I guess this might be an argument for actually not treating both cases the same. Here unreachable might arguably be fine because the behaviour we rely on is encapsulated within `stream_handler.rs`. For the other case, i.e., the one above where we return `RejectReason::Unknown` we'd have to rely that nobody changes `push_input` in a way that we don't expect.

Reply:
Given that it's all in a single `impl` I think it's passable. I mean we could also type `induct_message_impl` more precisely (e.g., return a sum type `(Response * CanisterMigrating) + (Request * (CanisterMigrating + CanisterNotFound + ...))`) but maybe an overkill here.

Reply:
AFAICT you can't do that in Rust as of now. Maybe I'm not up to date here?

I read a while ago that it is an often requested feature to have individual variants as their own type. Then perhaps you could return a kind of match statement. Something like
```Rust
fn my_fn() -> (_, RequestOrResponse::Request(_)) | (RejectReason::CanisterMigrating, RequestOrResponse::Response(_))`.
```
And then the compiler could use that in the `match` statement. For now, it seems we are stuck with just setting impossible combinations to `unreachable!()`.

Reply:
Yeah I was using just some made up syntax that'd need to become a bit more heavy weight in Rust. Concretely, you'd need `induct_message_impl` to return `Option<Either<(Response, ResponseErrors), (Request, RejectReason)>>` (where `Either` is basically the same thing as `Result`, just named differently since it's not supposed to distinguish between errors, just give you a sum). But reading that return type is a bit painful, and you'd also need to have an `enum ResponseErrors { CanisterMigrating }`, which is what I mean by it being heavy. 

But then you could match on the result of `induct_message_impl` without having any unreachable branches.